### PR TITLE
Implement snazzy new syntax parser

### PR DESCRIPTION
**Pull request by jnwatson**

### CODE REVIEW

**Comment by jnwatson** [comment]
Open question: whether to rename syntax2.py syntax.py. I sorta like it as a new file because it is way different.

File: synapse/tests/test_cortex.py [hunk 1]
```diff
@@ -1303,9 +1303,6 @@ def onAdd(node):
 
         async with self.getTestCoreAndProxy() as (realcore, core):
 
-            msgs = await alist(core.storm('|help'))
```
Comment (vEpiphyte) [inline synapse/tests/test_cortex.py]: is this now allowed? or is this now bad syntax?

Comment (jnwatson) [reply synapse/tests/test_cortex.py]: this is now bad syntax.

File: synapse/cortex.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 import synapse.lib.cache as s_cache
 import synapse.lib.layer as s_layer
 import synapse.lib.storm as s_storm
-import synapse.lib.syntax as s_syntax
+import synapse.lib.grammar as s_syntax
```
Comment (vEpiphyte) [inline synapse/cortex.py]: pyramid

File: synapse/datamodel.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
 import synapse.lib.coro as s_coro
 import synapse.lib.types as s_types
-import synapse.lib.syntax as s_syntax
+import synapse.lib.grammar as s_grammar
```
Comment (vEpiphyte) [inline synapse/datamodel.py]: alpha order

File: synapse/lib/grammar.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+StormCmdParser = lark.Lark(_grammar, start='stormcmdargs', propagate_positions=True)
+
+class Parser:
+    '''
+    Storm query parser
+    '''
+    def __init__(self, text, offs=0):
+
+        self.offs = offs
+        assert text is not None
```
Comment (vEpiphyte) [inline synapse/lib/grammar.py]: remove assert - .strip() will fail on the next call anyways

File: synapse/lib/grammar.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    def stormcmdargs(self):
+        '''
+        Parse command args that might have storm queries as arguments
+        '''
+        try:
+            tree = StormCmdParser.parse(self.text)
+        except lark.exceptions.LarkError as e:
+            raise self._larkToSynExc(e)
+        newtree = AstConverter(self.text).transform(tree)
+        assert isinstance(newtree, s_ast.Const)
```
Comment (vEpiphyte) [inline synapse/lib/grammar.py]: if this is a hard requirement lets move it away from being a `assert` and make it an `if` so it can't be compiled away

File: synapse/tests/test_cortex.py (hunk 1 above)
Comment (vEpiphyte) [reply synapse/tests/test_cortex.py]: we should add this as a test to throw exception somewhere if we don't have it already

File: synapse/tests/test_cortex.py [hunk 6]
```diff
@@ -2471,6 +2470,37 @@ class Bawk:
...
+            await _test('$(1 + 2 * 4)', 9)
+            await _test('$((1 + 2) * 4)', 12)
+            await _test('$(1 < 1)', 0)
+            await _test('$(1 <= 1)', 1)
+            await _test('$(1 > 1)', 0)
+            await _test('$(1 >= 1)', 1)
+            await _test('$(1 >= 1 + 1)', 0)
+            await _test('$(1 >= 1 + 1 * -2)', 1)
+            await _test('$(1 - 1 - 1)', -1)
+            await _test('$(4 / 2 / 2)', 1)
```
Comment (vEpiphyte) [inline synapse/tests/test_cortex.py]: what about 1/2?
what about 1/0?

File: synapse/tests/test_lib_grammar.py [hunk 7]
```diff
@@ -0,0 +1,1010 @@
+import unittest
+
+import synapse.exc as s_exc
+
+import synapse.lib.datfile as s_datfile
+import synapse.lib.grammar as s_grammar
+
+import synapse.tests.utils as s_t_utils
+import lark  # type: ignore
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_grammar.py]: import lark after unittest

File: synapse/tests/test_lib_grammar.py [hunk 8]
```diff
@@ -0,0 +1,1010 @@
...
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: foo], PivotInFrom: [AbsProp: edge:has], isjoin=False]',
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: foo], CmdOper: [Const: delnode, Const: ()]]',
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: foobar], FormPivot: [AbsProp: edge:has], isjoin=True]',
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: foobar], FormPivot: [AbsProp: edge:has], isjoin=False, PivotInFrom: [AbsProp: test:str], isjoin=True]',
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: foobar], FormPivot: [AbsProp: edge:has], isjoin=False, PivotInFrom: [AbsProp: test:str], isjoin=False]',
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: hello], EditPropSet: [RelProp: tick, Const: 2001]]',
+    'Query: [LiftPropBy: [Const: test:str, Const: =, Const: hello], EditPropSet: [RelProp: tick, Const: 2002]]',
+    "Query: [LiftPropBy: [Const: test:str, Const: =, Const: pennywise], CmdOper: [Const: noderefs, Const: ('--join', '-d', '9', '--traverse-edge')]]",
+    "Query: [LiftPropBy: [Const: test:
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_grammar.py]: Can this test be removed?

File: synapse/lib/grammar.py (hunk 5 above)
Comment (jnwatson) [reply synapse/lib/grammar.py]: it is a logic error of catastrophic consequence if this fails

File: synapse/tests/test_lib_grammar.py (hunk 8 above)
Comment (jnwatson) [reply synapse/tests/test_lib_grammar.py]: ok

File: synapse/tests/test_cortex.py (hunk 6 above)
Comment (jnwatson) [reply synapse/tests/test_cortex.py]: 1/0 raises an exception. Not sure if that's the right thing to do or not.

File: synapse/tools/storm_format.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
+# pragma: no cover
```
Comment (invisig0th) [inline synapse/tools/storm_format.py]: Not sure we want this as an independant tool.  Seems a bit special purpose and inconvenient.  Is this something we should integrate into cmdr functionality?

Comment (jnwatson) [reply synapse/tools/storm_format.py]: yes it is

File: synapse/tests/test_cortex.py (hunk 6 above)
Comment (jnwatson) [reply synapse/tests/test_cortex.py]: Deferring this. We need a generic way of moving along.

File: synapse/tools/storm_format.py (hunk 9 above)
Comment (jnwatson) [reply synapse/tools/storm_format.py]: Moved to lib

File: docs/synapse/userguides/storm_ref_filter.ipynb [hunk 10]
```diff
@@ -125,7 +125,7 @@
    "source": [
     "# Use previous temp cortex, define and print test query\n",
     "q = '<query> '\n",
-    "q1 = 'inet:fqdn'\n",
+    "q1 = 'inet:fqdn '\n",
```
Comment (invisig0th) [inline docs/synapse/userguides/storm_ref_filter.ipynb]: what's the purpose of this change?

File: synapse/tests/test_cortex.py (hunk 1 above)
Comment (invisig0th) [reply synapse/tests/test_cortex.py]: Can we put this back after the previous discussion and tweaks?

File: docs/synapse/userguides/storm_ref_filter.ipynb (hunk 10 above)
Comment (jnwatson) [reply docs/synapse/userguides/storm_ref_filter.ipynb]: I commented on it in chat a while back. The resulting query before was `inet:fqdn-inet:fqdn=google.com` which we decided was invalid.
